### PR TITLE
Include SOR when calculating IR

### PR DESCRIPTION
**Pull request by benedikt-voelkel**
* yields minimal fix to "timestamp out of bounds" when timestamp == SOR

* change "<" --> "<="

* add some more comments

### CODE REVIEW

**Comment by benedikt-voelkel** [comment]
@chiarazampolli @noferini @sawenzel 

I could change this (https://github.com/AliceO2Group/AliceO2/pull/12368/files#diff-6822b634d1d807fc76a798a4fb12534fe4c43885ed06dcad5696de79a1fef7d9R752) to something like
```cpp
(nextindex - 1, nextindex + 1)
```
if that is supposed to be more stable.

File: DataFormats/Detectors/CTP/src/Scalers.cxx [hunk 1]
```diff
@@ -744,9 +747,15 @@ std::pair<double, double> CTPRunScalers::getRateGivenT(double timestamp, int cla
...
-
+  } else if (nextindex == 1 || nextindex == mScalerRecordO2.size() - 1) {
+    // this happens if
+    // 1. there are only 2 ScalerRecord OR
+    // 2. the timestamp falls into the very first ScalerRecord OR
+    // 3. the timestamp falls in the second last ScalerRecord
     return std::make_pair(/*global mean rate*/ calcRate(0, mScalerRecordO2.size() - 1), /* current rate */ calcRate(nextindex - 1, nextindex));
+  } else {
+    // the timestamp falls somewhere else and the size of mScalerRecordO2 is at least 3
+    return std::make_pair(/*global mean rate*/ calcRate(0, mScalerRecordO2.size() - 1), /* current rate */ calcRate(nextindex - 1, nextindex + 1));
```
Comment (benedikt-voelkel) [inline DataFormats/Detectors/CTP/src/Scalers.cxx]: That would now sample "3 points".

File: DataFormats/Detectors/CTP/src/Scalers.cxx [hunk 2]
```diff
@@ -641,7 +641,7 @@ std::pair<double, double> CTPRunScalers::getRate(uint32_t orbit, int classindex,
   // orbit numbers
 
   // then we can use binary search to find the right entries
-  auto iter = std::lower_bound(mScalerRecordO2.begin(), mScalerRecordO2.end(), orbit, [&](CTPScalerRecordO2 const& a, uint32_t value) { return a.intRecord.orbit < value; });
+  auto iter = std::lower_bound(mScalerRecordO2.begin(), mScalerRecordO2.end(), orbit, [&](CTPScalerRecordO2 const& a, uint32_t value) { return a.intRecord.orbit <= value; });
```
Comment (benedikt-voelkel) [inline DataFormats/Detectors/CTP/src/Scalers.cxx:644]: Applying the same logic when retrieving the IR via orbit

**Comment by sawenzel** [comment]
> @chiarazampolli @noferini @sawenzel
> 
> I could change this (https://github.com/AliceO2Group/AliceO2/pull/12368/files#diff-6822b634d1d807fc76a798a4fb12534fe4c43885ed06dcad5696de79a1fef7d9R752) to something like
> 
> ```c++
> (nextindex - 1, nextindex + 1)
> ```
> 
> if that is supposed to be more stable.

I don't immediately see the need for this. Would fix real bugs first.

**Comment by benedikt-voelkel** [comment]
I leave that decision up to you and @chiarazampolli and @noferini .

The first commit is, as I indicated, the minimal solution to include `timestamp == SOR`.

**Comment by noferini** [comment]
Hi @benedikt-voelkel , have you tested it?
I can cherry-pick the change and test as well. Let me know

**Comment by benedikt-voelkel** [comment]
Yes, I am testing it right now.
Before any cherry picking, I would propose that we resolve @sawenzel 's comment first. Meaning
1. Should we go with the minimal solution (solving `timestamp == SOR`)
1. Should we in addition widen the range when deriving the local IR?

And then merge it here.

As I said, at this point, I don't have a feeling of what to do.
* Are the large fluctuations that are smoothened when we include a wider range for the IR? Is that quantified?
* Is there another reason why we should go with a wider range?

I cannot answer these questions with confidence.

**Comment by chiarazampolli** [comment]
Hello,

The existing issue comes from the fact that we use two points now, as far as i have understood. The rest of my comment is based on this assumption. 

The closest to instantaneous is to take only two points (one in fact would be best, if we knew precisely on how long it is integrated), so there should be a fix for start of run (was it checked that all failing jobs are at start of run? If yes, why 30% are at start of run? Was this maybe a very very short run?). This fix could take the first and second entry in the scalers, instead of "current" and "current - 1".

I have never seen large fluctuations, CTP itself averages over 10 seconds, but at the end of run. But these are then real drops, eg at beam dump. So, let's keep 2, since there are doubts. 

Chiara

File: DataFormats/Detectors/CTP/src/Scalers.cxx [hunk 3]
```diff
@@ -702,8 +702,11 @@ std::pair<double, double> CTPRunScalers::getRateGivenT(double timestamp, int cla
   // orbit numbers
 
   // then we can use binary search to find the right entries
-  auto iter = std::lower_bound(mScalerRecordO2.begin(), mScalerRecordO2.end(), timestamp, [&](CTPScalerRecordO2 const& a, uint32_t value) { return a.epochTime < value; });
-  auto nextindex = iter - mScalerRecordO2.begin(); // this points to the first index that has orbit greater or equal to given orbit
+  auto iter = std::lower_bound(mScalerRecordO2.begin(), mScalerRecordO2.end(), timestamp, [&](CTPScalerRecordO2 const& a, double value) { return a.epochTime <= value; });
```
Comment (benedikt-voelkel) [inline DataFormats/Detectors/CTP/src/Scalers.cxx:705]: @sawenzel @chiarazampolli  @noferini 
 It just occurred to me, there was previously `uint32_t` in the internal sample function of `getRateGivenT`.
That was implicitly casting the `double` timestamp to `uint32_t` which broke the comparison.
I corrected that. Apart from the `SOR` feature, that should solve some other issues we see.

**Comment by benedikt-voelkel** [comment]
In the test I am running (https://gitlab.cern.ch/o2simulationtools/alice-dpg-tests/-/tree/main/reco-2022/20231122?ref_type=heads) with run number `526467` the sampled time stamp from `o2dpg_sim_workflow_anchored.py` is 
`1664555124`
For that same run, the first Scaler has `epochTime == 1664555192`.
*NOW*: `1664555192 > 1664555124`.

Is there a problem of how we sample the final time?
https://github.com/AliceO2Group/O2DPG/blob/master/MC/bin/o2dpg_sim_workflow_anchored.py#L321

Anchor parameters used:
Run number: 526467
PRODSPLIT=15
SPLITID=1
CYCLE=0

**Comment by chiarazampolli** [comment]
> In the test I am running (https://gitlab.cern.ch/o2simulationtools/alice-dpg-tests/-/tree/main/reco-2022/20231122?ref_type=heads) with run number `526467` the sampled time stamp from `o2dpg_sim_workflow_anchored.py` is `1664555124` For that same run, the first Scaler has `epochTime == 1664555192`. _NOW_: `1664555192 > 1664555124`.
> 
> Is there a problem of how we sample the final time? https://github.com/AliceO2Group/O2DPG/blob/master/MC/bin/o2dpg_sim_workflow_anchored.py#L321
> 
> Anchor parameters used: Run number: 526467 PRODSPLIT=15 SPLITID=1 CYCLE=0

According to RCT/Info/RunInformation CCDB, which is what is used to determine start and end of run, we have:

SOR: 1664555124634
EOR: 1664558184117

Your timestamp seems in seconds, while it should be in ms (this can make a difference), but it is close to SOR. 
From our spreadsheet https://docs.google.com/spreadsheets/d/14vXFYVx3oVE8wgJKmIBhjvAt6NpGp7D6H4AmBM9E0Cw/edit#gid=1454168024, I see:

SOX: 1664555192000	
EOX: 1664558184000

So SOX is compatible. 
This is one example for which we should indeed generate using SOX and EOX, as I was saying in my previous comment. In principle, SOR < SOX and in general it should have been also EOR > EOX, but this second one was found not to be always true in 2023, and it is being discussed how to fix it. But to do the right thing, one should use SOX and EOX. 

Unfortunately SOX/EOX are not there for everything, and this ticket https://alice.its.cern.ch/jira/browse/O2-3868 should fix it. Maybe @noferini can take a look, and we do the upload and then we change and use SOX and EOX? 

The other option is that when the timestamp is < SOX, we use SOX from the scalers, and if > EOX, we use EOX from the scalers. This could be a quick fix. 

This is fine, and it would also fix the cases when the duration of the run in terms of triggers differed significantly from the duration using SOR and EOR, for which the sampling would be not 100% correct.

Does it make sense?

Chiara

**Comment by benedikt-voelkel** [comment]
Thanks for the explanation!
Since in the python script we retrieve both by run number, `RCT/Info/RunInformation` as well as `CTP/Calib/Scalers`, we could basically go for
`adjustedSOR = max(SOR, SOX)`
where we take `SOR` from `RCT/Info/RunInformation` and `SOX` from the first `CTP/Calib/Scalers`.
Of course we could simply always go for `SOX` from the Scalers, that would always give us a good timestamp since it anyway is expected to `SOX > SOR` as you say.

Then, when things are fully updated (https://alice.its.cern.ch/jira/browse/O2-3868), we can work entirely with `RCT/Info/RunInformation`.

Does that sound reasonable?

**Comment by chiarazampolli** [comment]
I think i would do like that, yes. Thanks!

**Pull request merged**